config/src/main/java/com/alibaba/nacos/config/server/service/notify/AsyncNotifyService.java

```diff
@@ -212,101 +357,61 @@ public void onCancel() {
-                        + "?dataId={2}&group={3}&tenant={4}";
-        
-        private int failCount;
-        
-        public NotifySingleTask(String dataId, String group, String tenant, long lastModified, String target) {
-            this(dataId, group, tenant, lastModified, target, false);
-        }
-        
-        public NotifySingleTask(String dataId, String group, String tenant, long lastModified, String target,
-                boolean isBeta) {
-            this(dataId, group, tenant, null, lastModified, target, isBeta);
-        }
-        
-        public NotifySingleTask(String dataId, String group, String tenant, String tag, long lastModified,
-                String target, boolean isBeta) {
-            super(dataId, group, tenant, lastModified);
-            this.target = target;
-            this.isBeta = isBeta;
-            try {
-                dataId = URLEncoder.encode(dataId, Constants.ENCODE);
-                group = URLEncoder.encode(group, Constants.ENCODE);
-            } catch (UnsupportedEncodingException e) {
-                LOGGER.error("URLEncoder encode error", e);
+        private void addTask(NotifySingleTask task) {
+            //sign the task is retry task. When it execute, just request once http invoke.
+            if (!task.isRetry()) {
+                task.setRetry(true);
+                task.initStartTimestamp();
             }
-            if (StringUtils.isBlank(tenant)) {
-                this.url = MessageFormat.format(URL_PATTERN, target, EnvUtil.getContextPath(), dataId, group);
-            } else {
-                this.url = MessageFormat
-                        .format(URL_PATTERN_TENANT, target, EnvUtil.getContextPath(), dataId, group, tenant);
-            }
-            if (StringUtils.isNotEmpty(tag)) {
-                url = url + "&tag=" + tag;
+            
+            NotifySingleTask existTask = delayNotifyTaskMap.putIfAbsent(task.getSingleTaskKey(), task);
+            //if the task exist, should extend it's start timestamp.
+            if (existTask != null) {
+                existTask.incFailCount();
+                existTask.resetStartTimestamp();
             }
-            failCount = 0;
-            // this.executor = executor;
         }
         
-        @Override
-        public void setFailCount(int count) {
-            this.failCount = count;
+        private void removeTask(NotifySingleTask task) {
+            delayNotifyTaskMap.remove(task.getSingleTaskKey());
         }
         
         @Override
-        public int getFailCount() {
-            return failCount;
+        public void run() {
+            long current = System.currentTimeMillis();
+            for (NotifySingleTask task : delayNotifyTaskMap.values()) {
+                if (current < task.getStartTimestamp()) {
+                    continue;
+                }
+                Queue<NotifySingleTask> queue = new LinkedList<>();
```

Comment:
Why can't the tasks here be saved into a List and submitted together

Reply:
If so, it just single thread to process it.